File: vscode-extension/src/services/git-service.ts
import * as vscode from 'vscode';
import { logger } from './logging-service';

export interface GitRepository {
    rootUri: vscode.Uri;
    state: {
        HEAD?: {
            commit?: string;
            name?: string;
        };
        onDidChange: (cb: () => void) => vscode.Disposable;
    };
}

export interface GitAPI {
    repositories: GitRepository[];
    onDidOpenRepository: (cb: (repo: GitRepository) => void) => vscode.Disposable;
}

export class GitService {
    private gitChangeDebounce: NodeJS.Timeout | undefined;
    private onRepoChange: () => Promise<void>;

    constructor(onRepoChange: () => Promise<void>) {
        this.onRepoChange = onRepoChange;
    }

    public async setupGitListener(context: vscode.ExtensionContext): Promise<void> {
        try {
            const gitExtension = vscode.extensions.getExtension('vscode.git');

            if (!gitExtension) {
                logger.log('Git extension not found - skipping git change detection');
                return;
            }

            if (!gitExtension.isActive) {
                await gitExtension.activate();
            }

            const gitApi = gitExtension.exports?.getAPI?.(1) as GitAPI | undefined;

            if (!gitApi || typeof gitApi !== 'object') {
                logger.error('Failed to get git API object');
                return;
            }

            logger.log('Git API obtained successfully');

            gitApi.repositories.forEach((repo) => this.setupRepositoryListener(repo));
            context.subscriptions.push(gitApi.onDidOpenRepository((repo) => this.setupRepositoryListener(repo)));

            logger.log(`Git listener setup complete. Monitoring ${gitApi.repositories.length} repositories.`);
        } catch (error) {
            logger.error('Error setting up git listener', error);
        }
    }

    private setupRepositoryListener(repository: GitRepository): void {
        if (!repository || typeof repository !== 'object') return;

        let lastHead = repository.state.HEAD?.commit || repository.state.HEAD?.name;
        logger.log(`Monitoring repository at ${repository.rootUri.fsPath} (Initial HEAD: ${lastHead})`);

        repository.state.onDidChange(() => {
            const currentHead = repository.state.HEAD?.commit || repository.state.HEAD?.name;

            if (currentHead !== lastHead) {
                logger.log(`Git detected HEAD change in ${repository.rootUri.fsPath}: ${lastHead} -> ${currentHead}`);
                lastHead = currentHead;

                if (this.gitChangeDebounce) {
                    clearTimeout(this.gitChangeDebounce);
                }

                this.gitChangeDebounce = setTimeout(async () => {
                    logger.log(
                        `[Git Event] Head moved from ${lastHead} to ${currentHead}. Triggering full workspace refresh.`,
                    );
                    await this.onRepoChange();
                    logger.log('[Git Event] Workspace refresh finished.');
                }, 3000);
            }
        });
    }

    public dispose(): void {
        if (this.gitChangeDebounce) {
            clearTimeout(this.gitChangeDebounce);
        }
    }
}

File: visual-studio-extension/DeepLensVisualStudio/DeepLensVisualStudio/Services/GitService.cs
using System.Diagnostics;
using Microsoft.VisualStudio.Shell;

namespace DeepLensVisualStudio.Services
{
    /// <summary>
    /// Service that monitors git repository changes and triggers re-indexing on branch switches.
    /// </summary>
    public class GitService : IDisposable
    {
        private readonly Func<Task> _onRepoChange;
        private Timer? _debounceTimer;
        private string? _lastHead;
        private bool _disposed;

        public GitService(Func<Task> onRepoChange)
        {
            _onRepoChange = onRepoChange ?? throw new ArgumentNullException(nameof(onRepoChange));
        }

        public async Task SetupGitListenerAsync()
        {
            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

            try
            {
                // Try to get Git Extensibility service via reflection (may not be available)
                var gitExtType = Type.GetType("Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt, Microsoft.VisualStudio.TeamFoundation.Git.Extensibility");
                if (gitExtType == null)
                {
                    Debug.WriteLine("DeepLens: Git extensibility API not available - skipping git change detection");
                    return;
                }

                var gitExt = ServiceProvider.GlobalProvider.GetService(gitExtType);
                if (gitExt == null)
                {
                    Debug.WriteLine("DeepLens: Git extension not found - skipping git change detection");
                    return;
                }

                // Use reflection to access Git API methods
                var getRepoInfoMethod = gitExtType.GetMethod("GetRepositoryInfo");
                if (getRepoInfoMethod != null)
                {
                    var repoInfo = getRepoInfoMethod.Invoke(gitExt, null);
                    if (repoInfo != null)
                    {
                        var stateChangedEvent = repoInfo.GetType().GetEvent("StateChanged");
                        if (stateChangedEvent != null)
                        {
                            var handler = Delegate.CreateDelegate(stateChangedEvent.EventHandlerType!, this, nameof(OnGitStateChanged));
                            stateChangedEvent.AddEventHandler(repoInfo, handler);
                        }

                        var currentBranchProp = repoInfo.GetType().GetProperty("CurrentBranch");
                        if (currentBranchProp != null)
                        {
                            _lastHead = currentBranchProp.GetValue(repoInfo) as string;
                            Debug.WriteLine($"DeepLens: Monitoring git repository. Initial HEAD: {_lastHead}");
                        }
                    }
                }

                Debug.WriteLine("DeepLens: Git listener setup complete");
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"DeepLens: Error setting up git listener: {ex.Message}");
            }
        }

        private void OnGitStateChanged(object? sender, EventArgs e)
        {
            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
            {
                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

                try
                {
                    var gitExtType = Type.GetType("Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt, Microsoft.VisualStudio.TeamFoundation.Git.Extensibility");
                    if (gitExtType == null) return;

                    var gitExt = ServiceProvider.GlobalProvider.GetService(gitExtType);
                    if (gitExt == null) return;

                    var getRepoInfoMethod = gitExtType.GetMethod("GetRepositoryInfo");
                    if (getRepoInfoMethod == null) return;

                    var repoInfo = getRepoInfoMethod.Invoke(gitExt, null);
                    if (repoInfo == null) return;

                    var currentBranchProp = repoInfo.GetType().GetProperty("CurrentBranch");
                    if (currentBranchProp == null) return;

                    var currentHead = currentBranchProp.GetValue(repoInfo) as string;

                    if (currentHead != _lastHead)
                    {
                        Debug.WriteLine($"DeepLens: Git detected HEAD change: {_lastHead} -> {currentHead}");
                        _lastHead = currentHead;

                        // Debounce updates (3 second delay)
                        _debounceTimer?.Dispose();
                        _debounceTimer = new Timer(async _ =>
                        {
                            Debug.WriteLine($"[Git Event] Head moved from {_lastHead} to {currentHead}. Triggering full workspace refresh.");
                            await _onRepoChange();
                            Debug.WriteLine("[Git Event] Workspace refresh finished.");
                        }, null, TimeSpan.FromSeconds(3), Timeout.InfiniteTimeSpan);
                    }
                }
                catch (Exception ex)
                {
                    Debug.WriteLine($"DeepLens: Error in git state change handler: {ex.Message}");
                }
            });
        }

        public void Dispose()
        {
            if (_disposed) return;

            _debounceTimer?.Dispose();
            _debounceTimer = null;

            try
            {
                var gitExtType = Type.GetType("Microsoft.VisualStudio.TeamFoundation.Git.Extensibility.IGitExt, Microsoft.VisualStudio.TeamFoundation.Git.Extensibility");
                if (gitExtType != null)
                {
                    var gitExt = ServiceProvider.GlobalProvider.GetService(gitExtType);
                    if (gitExt != null)
                    {
                        var getRepoInfoMethod = gitExtType.GetMethod("GetRepositoryInfo");
                        if (getRepoInfoMethod != null)
                        {
                            var repoInfo = getRepoInfoMethod.Invoke(gitExt, null);
                            if (repoInfo != null)
                            {
                                var stateChangedEvent = repoInfo.GetType().GetEvent("StateChanged");
                                if (stateChangedEvent != null)
                                {
                                    var handler = Delegate.CreateDelegate(stateChangedEvent.EventHandlerType!, this, nameof(OnGitStateChanged));
                                    stateChangedEvent.RemoveEventHandler(repoInfo, handler);
                                }
                            }
                        }
                    }
                }
            }
            catch { }

            _disposed = true;
        }
    }
}
